staging/src/k8s.io/apiserver/pkg/admission/chain.go

```diff
@@ -16,22 +16,45 @@ limitations under the License.
 
 package admission
 
-// chainAdmissionHandler is an instance of admission.Interface that performs admission control using a chain of admission handlers
-type chainAdmissionHandler []Interface
+import "time"
+
+// chainAdmissionHandler is an instance of admission.Interface that performs admission control using
+// a chain of admission handlers
+type chainAdmissionHandler []NamedHandler
 
 // NewChainHandler creates a new chain handler from an array of handlers. Used for testing.
-func NewChainHandler(handlers ...Interface) chainAdmissionHandler {
+func NewChainHandler(handlers ...NamedHandler) chainAdmissionHandler {
 	return chainAdmissionHandler(handlers)
 }
 
+func NewNamedHandler(name string, i Interface) NamedHandler {
+	return &pluginHandler{
+		i:    i,
+		name: name,
+	}
+}
+
+const (
+	stepValidating = "validating"
+	stepMutating   = "mutating"
+)
+
 // Admit performs an admission control check using a chain of handlers, and returns immediately on first error
 func (admissionHandler chainAdmissionHandler) Admit(a Attributes) error {
+	var err error
```

Comment:
nit: this is fragile since it's easy to accidentally mask this with
```
err := foo()
return err
```
But if you make it a named return instead, you don't need to worry about that, e.g. `Admit(a Attributes) (err error) { ...`

Same applies below (Validate)

Reply:
I don't follow this suggestion. the only difference between starting off with `var err error` and using a named return parameter is that in the former case, empty `return` won't work? Either way it can be shadowed in a nested {} block.

Using a named parameter saves a line of code.

If anything, it is the entire pattern of maintaining state in a variable throughout the entire function to return it at the end that is suspect. I believe the purpose is to defer the Observe call; but that could be handled differently:

```go
finish := func(err error) error {
  Metrics.ObserveAdmissionStep(time.Since(start), err != nil, a, stepMutating)
  return err
}

// And then instead of returning an error directly, always:
return finish(err)

```

Reply:
It gets masked in the scope of the variable, but the defered function still sees the return result: https://gist.github.com/tallclair/b2e6d2217391cbfeafd919bb2f842b46

Reply:
Good catch @tallclair 

How about splitting this into two functions to avoid maintiaining the state throughout the entire function?

```go
func (admissionHandler chainAdmissionHandler) Validate(a Attributes) error {
     start := time.Now()
     err := admissionHandler.validate(a)
     Metrics.ObserveAdmissionStep(time.Since(start), err != nil, a, stepValidating)
     return err
 }

 func (admissionHandler chainAdmissionHandler) validate(a Attributes) (err error) {
    // original, uninstrumented implementation
  }
```

Reply:
works for me